Feature/pem to key

owljoa opened the pull request:
찾아보니 GetKey 라는 함수가 keyManager에 있었습니다. :)
@junbeomlee GetKey 함수를 이용하면 key file path에 있는 file 형태의 key를 PEM -> DER 포멧을 거쳐서 -> Key형태로 반환됩니다.
(GetKey에서 파일명을 복원하는데 중간에 keyGenOpts에 해당하는 부분이 빠져있어서 동작을 제대로 못하고 있는것을 수정했습니다.)
말씀하신 algorithm, type, byte를 받아 key를 반환하는 함수는.. GetKey이외에도 저런 형태의 함수가 꼭 필요하시다면 다시 한번 말씀해주세요! :)

stilere committed:
add keyGenOpt parameter to loadKey function for fixing error of
keyManager.GetKey function.

stilere committed:
add GetPath in keyManager for check the path of key file on main
package.

- sample files changed using GetPath.
- now, using the sample files, we can get key from default key directory
if key already stored in the directory.

yojkim reviewed (approved):
아 요런 실수를 ^^7; 감사합니다.

stilere committed:
add ByteToKey function to keyManager.

- add ByteToKey function
- modify comment of GetKey function

stilere committed:
add ByteToKey function test code into sample files.

owljoa commented:
추가 요청하신 기능이 이 브랜치에 속하는것 같아 여기서 작업했습니다!
keyManager에 ByteToKey라는 이름의 함수를 추가했고 sample 파일들에도 관련 코드 추가했습니다 :)

혹시 오류나 잘못된 부분이 있다면 알려주세요!

yojkim commented on sample/heimdall-ecdsa-sample.go:
```diff
@@ -24,15 +26,31 @@ func main() {
-	errorCheck(err)
-
 	// if there is no key file in default key directory, then generate key.
 	if _, err := os.Stat(keyPath); os.IsNotExist(err) {
```
저장 경로에 키가 있는지 체크해주는 로직을 `GenerateKey` 내부에 포함시키는 것은 어떨까요? 외부에서 해당 경로를 항상 체크하는게 뭔가 번거로울거 같아서욥!!
사용자가 키를 처음에 생성하고 키까지 바로 받을 수 있도록하면 `GetKey`를 따로 호출할 필요도 없을거같고
만약 이후에 별도로 키를 받아볼 상황이 필요하다면 그 때 `GetKey`를 호출해도 될 것 같은데 어떻게 생각하세요?

owljoa replied:
그 로직은 generateKey 내부에서 호출되는 storeKey 함수에 있는 것 같아요!
sample 소스는 제가 급히 짜다보니 불필요하게 복잡해진 것 같기도 하네요 하하;

yojkim replied:
아하 그러네여 ㅋㅋㅋㅋㅋ 제가짜고도 헷갈린;;;

junbeomlee commented:
로직을 보니 https://github.com/it-chain/heimdall/blob/master/key/keyLoader.go 여기서 file을 읽는 부분하고 중복되는 부분들이 많은 것 같습니다! 중복을 줄일 수 있도록 최대한 코드를 재사용하는 방향으로 구현하면 더 좋은 코드가 나올 수 있을 것 같습니다.

stilere committed:
Merge branch 'master' of https://github.com/it-chain/heimdall into feature/pem_to_key

stilere committed:
add MatchPrivateKeyOpt, MatchPublicKeyOpt functions into keyUtil.go
file.

- Both functions convert key interface type to matched private or public
key type using key generation option.
- some comments for keyUtil.go file are added.

stilere committed:
use MatchPrivateKeyOpt, MatchPublicKeyOpt for converting key interface type
to matched private or public key type.

- replace original code with the two functions.

stilere committed:
use MatchPrivateKeyOpt, MatchPublicKeyOpt functions for converting key
interface type to matched private or public key type.

- replace original codes with the two functions.

stilere committed:
temp commit for prototyping the MatchPrivateKeyOpt, MatchPublicKeyOpt
functions.

owljoa commented:
중복된 코드들을 keyUtil.go에 함수로 만들었습니다!

key interface 타입을 keyGenOpt에 맞춰서 해당하는(rsa or ecdsa) 알고리즘의 private key 타입으로 변환하는 함수와 해당하는 알고리즘의 public key 타입으로 변환하는 함수 두개로 만들었습니다.

함수이름은 일단 MatchPrivateKeyOpt, MatchPublicKeyOpt로 지었는데.. ㅠ 뭔가 안맞는것 같아서 조언 부탁드립니다..!

그리고 sample 코드는 일단 기능 확인하시기 편하게 rsa 샘플코드에서 GetKey, ByteToKey 두가지에 집중했습니다.

sample 코드는 내일 알맞게 수정하고, usage도 함께 수정하겠습니다 :)

yojkim commented:
좋은 것 같아요!!!
